### PR TITLE
#8600: drop deprecated `localStorage` data binding for embedded forms

### LINKED ISSUE #8600: Drop deprecated data location from embedded form
Context
---
- Use package search to verify not used in production mods: `$.*[?(@.type=="localStorage")]`
- https://pixiebrix.slack.com/archives/C02RAMRDF0R/p1718130254506609

### PR DESCRIPTION
**Pull request by twschiller**
## What does this PR do?

- Closes #8600 
- Drops deprecated `localStorage` data binding, including dropping the default value passed for "storage"
- Makes `storage` prop required on custom form
- Rewrite tests to use mod variables

## Discussion

- The change is backward incompatible. The field had shown as Deprecated in the Page Editor for a long time
- Will break the inline panel in test mod `@pixies/linkedin/organization-demos` which doesn't declare a storage type (that mod was hand-written in v1). We could switch that test mod to use mod variable, or just drop that part of the test. (I'd vote we just drop that part of the mod given we're not supporting that use case)
 
## Customer Impact Checks for Backward Compatibility

✅  Query: `$..*[?(@.type=="localStorage")]`

No customer mods impacted

## Future Work

- There's some future improvements we'll need to make. Custom form doesn't use the Platform API for mod variables, it calls the content script messenger directly

## Checklist

- [x] Add jest or playwright tests and/or storybook stories
- [x] Designate a primary reviewer: @fungairino

### CODE REVIEW

File: src/bricks/renderers/customForm.tsx [hunk 1]
```diff
@@ -398,17 +393,16 @@ async function getInitialData(
...
       const topLevelFrame = await getConnectedTarget();
-      // Target the top level frame. Inline panels aren't generally available, so the renderer will always be in the
-      // sidebar which runs in the context of the top-level frame
+      // XXX: CustomForm currently uses page state directly instead of the platform's state API. Switching to the
+      //  platform's state API would be a breaking change for any forms that are relying on the top-level frame behavior
+      // Original comment:
+      //  Target the top level frame. Inline panels aren't generally available, so the renderer will always be in the
+      //  sidebar which runs in the context of the top-level frame
+      // The above isn't true anymore, because custom form could appear in a modal or popover, which could be in the
+      // context of a frame
```
Comment (BLoe) [inline src/bricks/renderers/customForm.tsx:405]: Is it possible to "resolve" the conflicting points in these comments somehow? Should we just remove the old comment? It feels like this is just getting really confusing reading this back as another engineer without all the context.

File: src/bricks/renderers/customForm.tsx [hunk 2]
```diff
@@ -165,7 +153,7 @@ export const CUSTOM_FORM_SCHEMA: Schema = {
     recordId: {
       type: "string",
       description:
-        "Unique identifier for the data record. Required if using a database or local storage",
+        "Unique identifier for the data record. Required if using a database storage",
```
[the remark this answers is not in the document]
Comment (fungairino) [reply src/bricks/renderers/customForm.tsx:156]: +1 to "using a database"

**Comment by fungairino** [comment]
Learning:

How did you make the query you mentioned in the description?
"$..*[?(@.type=="localStorage")]"

File: src/bricks/renderers/customForm.tsx (hunk 1 above)
Comment (twschiller) [reply src/bricks/renderers/customForm.tsx]: @BLoe @fungairino see updated text

**Comment by twschiller** [comment]
> Learning:
> 
> How did you make the query you mentioned in the description? `"$..*[?(@.type=="localStorage")]"`

See https://app.pixiebrix.com/registry-query which has a linked Notion with examples. I had to use command line because the query was taking longer than 30s

File: src/bricks/renderers/customForm.tsx [hunk 3]
```diff
@@ -63,7 +62,8 @@ export type StateStorage = {
 };
 
 export type Storage =
-  | { type: "localStorage" }
+  // Support for `localStorage` was removed in 2.0.3
+  // | { type: "localStorage"}
```
Comment (fungairino) [inline src/bricks/renderers/customForm.tsx:66]: Can we remove unused commented out code? If we need it as a reference maybe we can move this note to a jsdoc for the Storage type?

File: src/pageEditor/fields/FormRendererOptions.tsx [hunk 4]
```diff
@@ -168,7 +167,7 @@ const FormDataBindingOptions: React.FC<{
         />
       )}
 
-      {["localStorage", "database"].includes(storageType) && (
+      {["database"].includes(storageType) && (
```
Comment (fungairino) [inline src/pageEditor/fields/FormRendererOptions.tsx:170]: nit: simplify to
```suggestion
     storageType === "database" && (
```

**Review by fungairino — APPROVED** [review]

**Pull request merged**